src/os/bluestore/BlueStore.h

```diff
@@ -1289,7 +1289,7 @@ class BlueStore : public ObjectStore,
 
   private:
     /// forward lookups
-    mempool::bluestore_cache_meta::unordered_map<ghobject_t,OnodeRef> onode_map;
+    mempool::bluestore_cache_meta::map<ghobject_t,OnodeRef> onode_map;
```

Comment:
it looks like you're using a single hash for all the objects in your benchmark in https://github.com/ceph/ceph/pull/39976
May be that's the key of bad performance for unordered_map?

Reply:
We might also provide non-default bucket count when constructing unordered_map. AFAIK default one is 1 which results in multiple rehashings when doing bulk inserting... But this is rather not a big win anyway...

Reply:
> it looks like you're using a single hash for all the objects in your benchmark in #39976
> May be that's the key of bad performance for unordered_map?

Sorry, I haven't updated that PR yet.  My local version lets you configure the number of unique hashes to use during the benchmark.  I also verified that unique hobject_t hashes have a 1:1 mapping with std::hash using the ghobject_t hash specialization.  In this case, there are 100 unique hobject_t "hash" values and 100 unique ghobject_t hashes via the hash specialization.

Reply:
> We might also provide non-default bucket count when constructing unordered_map. AFAIK default one is 1 which results in multiple rehashings when doing bulk inserting... But this is rather not a big win anyway...

Do you mean increasing the capacity via reserve()?  I experimented with that but saw no benefit as you suspected.  The cost appears to be quickly amortized due to the semi-permanent nature of the onode_map.